File: dotnet/src/Pactman.NonprofitCheckPlus/Exceptions/PactmanTimeoutException.cs
using System;
using System.Collections.Generic;

using Pactman.NonprofitCheckPlus.Internal;

namespace Pactman.NonprofitCheckPlus.Exceptions
{
    /// <summary>
    /// The request exceeded the configured timeout.
    /// </summary>
    /// <remarks>
    /// Raised only when the SDK's own deadline expired. A <see cref="System.Threading.CancellationToken"/>
    /// you cancelled surfaces as <see cref="OperationCanceledException"/> instead, so
    /// "I gave up" stays distinguishable from "the API was too slow".
    /// </remarks>
    public sealed class PactmanTimeoutException : PactmanException
    {
        /// <summary>Initializes the exception.</summary>
        /// <param name="message">A description naming the deadline that elapsed.</param>
        /// <param name="timeout">The timeout that elapsed.</param>
        /// <param name="attempts">How many attempts were made before this error was surfaced.</param>
        /// <param name="innerException">The underlying cancellation or transport failure.</param>
        public PactmanTimeoutException(
            string message,
            TimeSpan timeout,
            int attempts = 1,
            Exception? innerException = null)
            : base(message, ErrorCategory.Timeout, ErrorOrigin.Local, innerException)
        {
            Timeout = timeout;
            Attempts = attempts;
        }

        /// <summary>The timeout that elapsed.</summary>
        public TimeSpan Timeout { get; }

        /// <summary>How many attempts were made before this error was surfaced.</summary>
        public int Attempts { get; }

        /// <inheritdoc />
        public override IReadOnlyDictionary<string, object?> ToDictionary() =>
            new FieldMap(base.ToDictionary())
            {
                ["timeout"] = Timeout.TotalSeconds,
                ["attempts"] = Attempts,
            };
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/exceptions/PactmanTimeoutException.java
package org.pactman.nonprofitcheckplus.exceptions;

import java.util.Map;

/** The request exceeded the configured timeout. */
public class PactmanTimeoutException extends PactmanException {

    private static final long serialVersionUID = 1L;

    /** @serial the timeout that was exceeded, in milliseconds. */
    private final long timeoutMs;

    /** @serial how many attempts were made. */
    private final int attempts;

    /**
     * Creates the exception.
     *
     * @param message   what timed out.
     * @param timeoutMs the timeout that was exceeded, in milliseconds.
     * @param attempts  how many attempts were made.
     * @param cause     the underlying failure, or {@code null}.
     */
    public PactmanTimeoutException(String message, long timeoutMs, int attempts, Throwable cause) {
        super(message, ErrorCategory.TIMEOUT, ErrorOrigin.LOCAL, cause);
        this.timeoutMs = timeoutMs;
        this.attempts = attempts;
    }

    /**
     * The timeout that was exceeded.
     *
     * @return the timeout in milliseconds.
     */
    public long timeoutMs() {
        return timeoutMs;
    }

    /**
     * How many attempts were made before this error was surfaced.
     *
     * @return the attempt count, counting the first attempt as one.
     */
    public int attempts() {
        return attempts;
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> fields = super.toMap();
        fields.put("timeoutMs", timeoutMs);
        fields.put("attempts", attempts);

        return fields;
    }
}
